backend/app/treasury/adapters/covalent/transfers_v2.py

```diff
@@ -10,30 +10,65 @@
 from .... import db
 from ...models import Transfer
 
-CACHE_HASH_TRANSFERS = "covalent_transfers"
+# [XXX] We might want to have one Redis key per token.
+# The big hashmap is probably not playing well with expiry.
+CACHE_HASH_TRANSFERS = "covalent_transfer_items"
 CACHE_KEY_TEMPLATE_TRANSFERS = "{treasury_address}_{contract_address}_{chain_id}_{date}"
 
+KEY = os.getenv("COVALENT_KEY")
+TRANSFERS_V2_URL_TEMPLATE = (
+    "https://api.covalenthq.com/v1/{chain_id}/address/{treasury_address}/transfers_v2/"
+)
 
-async def _get_transfers_data(
+
+async def _get_transfer_items(
     treasury_address: str, contract_address: str, chain_id: int
-) -> dict[str, Any]:
-    async with AsyncClient(timeout=Timeout(10.0, read=15.0, connect=30.0)) as client:
-        resp = await client.get(
-            f"https://api.covalenthq.com/v1/{chain_id}/address/{treasury_address}"
-            + "/transfers_v2/?quote-currency=USD&format=JSON"
-            + f"&contract-address={contract_address}&key=ckey_{getenv('COVALENT_KEY')}"
-        )
-    resp.raise_for_status()
-    return resp.json()["data"]
+) -> Generator[Any, None, None]:
+    page_number = 0
+    while True:
+        async with AsyncClient(
+            timeout=Timeout(10.0, read=60.0, connect=90.0)
+        ) as client:
+            resp = await client.get(
+                TRANSFERS_V2_URL_TEMPLATE.format(
+                    chain_id=chain_id, treasury_address=treasury_address
+                ),
+                params={
+                    "quote-currency": "USD",
+                    "format": "JSON",
+                    "contract-address": contract_address,
+                    "key": f"ckey_{KEY}",
+                    "page-number": page_number,
+                },
+            )
+        resp.raise_for_status()
+        data = resp.json()["data"]
+        for item in data["items"]:
+            yield item
+        if not data["pagination"]["has_more"]:
+            break
+        page_number += 1
 
 
 TYPE_SIGN = {"OUT": -1, "IN": 1}
 
 
+def _transfers_of_items(transfer_items: list[Any]):
+    for block_transaction in transfer_items:
+        block_date = dateutil.parser.parse(block_transaction["block_signed_at"])
+        for transfer in block_transaction["transfers"]:
+            delta = int(transfer["delta"])
+            decimals = int(transfer["contract_decimals"])
+            amount = (
+                TYPE_SIGN[transfer["transfer_type"]] * delta / 10**decimals
```

Comment:
The `0` case is well handled by `10**decimals`.

Have you encountered cases with decimals being negative?

It should never be the case according to [ERC20](https://ethereum.org/en/developers/docs/standards/tokens/erc-20/), as decimals is `uint8`, unsigned.

Thus, if ever Covalent has returned decimal values < 0, we should consider this piece of data to be corrupted. I see two candidate reasons:

1. A hack from the token developers, who have not followed ERC20 and rather have made `decimals()` return a signed integer, and a negative one. I believe these cases are well filtered-out by our whitelist mechanism (we can even verify this as an interesting exercise).
2. The data provider, Covalent, is wrong: they are returning a decimals value incoherent with the contract's `decimals()`. I consider this to be higly unlikely. More, if such a bug would ever happen on their end, I believe they would fix it after some time.

As a conclusion, let's consider Covalent's responses to be having decimals ≥ 0. I'm not in favor of adding a new clause to the code to handle such cases, to keep complexity in check.  
If we considered our data source to be potentially corrupt, to me the right solution would not be to fallback to decimals=0, but rather to build our own whitelisting mechanism. I believe we don't need to consider this now.

To respond to your concern that this may still happen (albeit with very low probability), I suggest to log an error whenever `decimals` is negative. That will bubble up in Sentry.

Reply:
Logging added.

Reply:
Fair points, I agree with the logging solution.